src/bundle.ts

```diff
@@ -0,0 +1,79 @@
+import * as path from "path";
+import { TranspiledFileWithSourceNode, EmitHost } from "./Transpile";
+import { SourceNode } from "source-map";
+import { formatPathToLuaPath, trimExt } from "./utils";
+import { Diagnostic } from "typescript";
+import { couldNotFindBundleEntryPoint } from "./diagnostics";
+
+const formatPath = (path: string) => formatPathToLuaPath(trimExt(path));
+
+export function bundleTranspiledFiles(
+    bundleFile: string,
+    entryModule: string,
+    transpiledFiles: TranspiledFileWithSourceNode[],
+    emitHost: EmitHost
+): [Diagnostic[], TranspiledFileWithSourceNode] {
+    const diagnostics: Diagnostic[] = [];
+
+    if (transpiledFiles.find(f => f.fileName === entryModule) === undefined) {
+        return [[couldNotFindBundleEntryPoint(entryModule)], { fileName: bundleFile }];
+    }
+
+    // For each file: ["<file name>"] = function() <lua content> end,
+    const moduleTableEntries: SourceChunk[] = transpiledFiles.map(moduleSourceNode);
+
+    // If any of the modules contains a require for lualib_bundle, add it to the module table.
+    const lualibRequired = transpiledFiles.some(f => f.lua && f.lua.match(/require\("lualib_bundle"\)/));
+    if (lualibRequired) {
+        const lualibBundle = emitHost.readFile(path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua"));
+        moduleTableEntries.push(`["lualib_bundle"] = function() ${lualibBundle} end,\n`);
+    }
+
+    // Create ____modules table containing all entries from moduleTableEntries
+    const moduleTable = createModuleTableNode(moduleTableEntries);
+
+    // Override `require` to read from ____modules table.
+    const requireOverride =
+        `function require(file) if ____modules[file] then return ____modules[file]() ` +
```

Comment:
We should either fall back to original `require` if module is not found, or use a different function name